File: neural_network_python/neural_network.py
from dense_layer import DenseLayer
from utils import ActFunc
import utils

class NeuralNetwork:

    # ------------------------------------------------------------------------------
    def __init__(self, 
                 numInputs: int, 
                 numHiddenNodes: int, 
                 numOutputs: int,
                 actFuncHidden: ActFunc = ActFunc.Relu,
                 actFuncOutput: ActFunc = ActFunc.Relu) -> None:
        self._hiddenLayer = DenseLayer(numHiddenNodes, numInputs, actFuncHidden)
        self._outputLayer = DenseLayer(numOutputs, numHiddenNodes, actFuncOutput)
        self._trainInput = []
        self._trainOutput = []
        self._trainOrder = []

    # ------------------------------------------------------------------------------ 
    def numTrainingSets(self) -> int:
        return len(self._trainOrder)

    # ------------------------------------------------------------------------------ 
    def output(self) -> tuple[float]:
        return self._outputLayer.output()

    # ------------------------------------------------------------------------------ 
    def addTrainingData(self, trainInput: tuple[tuple[float]], trainOutput: tuple[tuple[float]]) -> bool:  
        self._trainInput = utils.toContainer(trainInput)
        self._trainOutput = utils.toContainer(trainOutput)
        self._checkTrainingData()
        self._initTrainOrderList()
        return self.numTrainingSets() > 0

    # ------------------------------------------------------------------------------ 
    def train(self, numEpochs: int, learningRate: float = 0.01) -> bool:
        if self.numTrainingSets() == 0 or numEpochs <= 0 or learningRate <= 0: 
            return False
        for i in range(numEpochs):
            self._randomizeTrainingOrder()
            for j in self._trainOrder:
                input = utils.toContainer(self._trainInput[j])
                output = utils.toContainer(self._trainOutput[j])
                self._feedforward(input)
                self._backpropagate(output)
                self._optimize(input, learningRate)
        return True
    
    # ------------------------------------------------------------------------------ 
    def predict(self, input: tuple[float]) -> tuple[float]:
        self._feedforward(input)
        return self.output()
    
    # ------------------------------------------------------------------------------ 
    def printPredictions(self, inputSets: tuple[tuple[float]], numDecimals: int = 1) -> None:
        if len(inputSets) == 0 or numDecimals < 0: return
        print("--------------------------------------------------------------------------------")
        for input in inputSets:
            print("Input:", end="\t")
            utils.printFloats(input, numDecimals)
            print("Output:", end="\t")
            utils.printFloats(self.predict(input), numDecimals)  
            if (input != inputSets[len(inputSets) - 1]): print()
        print("--------------------------------------------------------------------------------\n")

    # ------------------------------------------------------------------------------ 
    def _checkTrainingData(self) -> None:
        if len(self._trainInput) != len(self._trainOutput):
            numSets = utils.min(len(self._trainInput), len(self._trainOutput))
            self._trainInput = self._trainInput[:numSets]
            self._trainOutput = self._trainOutput[:numSets]

    # ------------------------------------------------------------------------------ 
    def _initTrainOrderList(self) -> None:
        for i in range(len(self._trainInput)):
            self._trainOrder.append(i)

    # ------------------------------------------------------------------------------ 
    def _randomizeTrainingOrder(self) -> None:
        import random
        for i in range(self.numTrainingSets()):
            r = random.randint(0, len(self._trainOrder) - 1)
            temp = self._trainOrder[i]
            self._trainOrder[i] = self._trainOrder[r]
            self._trainOrder[r] = temp

    # ------------------------------------------------------------------------------ 
    def _feedforward(self, input: tuple[float]) -> None:
        self._hiddenLayer.feedforward(input)
        self._outputLayer.feedforward(self._hiddenLayer.output())

    # ------------------------------------------------------------------------------ 
    def _backpropagate(self, output: tuple[float]) -> None:
        self._outputLayer.backpropagateOutput(output)
        self._hiddenLayer.backpropagateHidden(self._outputLayer)

    # ------------------------------------------------------------------------------ 
    def _optimize(self, input: tuple[float], learningRate: float) -> None:
        self._hiddenLayer.optimize(input, learningRate)
        self._outputLayer.optimize(self._hiddenLayer.output(), learningRate)

File: neural_network_cpp/src/neural_network.cpp
#include <neural_network.hpp>

namespace {
    
// --------------------------------------------------------------------------------
template <typename T = int>
void Print(const std::vector<T>& data, std::ostream& ostream = std::cout) {
    static_assert(std::is_arithmetic<T>::value, 
        "Non-arithmetic type selected for method ::Print!");
    ostream << "[";
    for (const auto& i : data) {
        ostream << i;
        if (&i < &data[data.size() - 1]) { ostream << ", "; }
    }
    ostream << "]\n";
}
} // namespace

namespace yrgo {
namespace machine_learning {

// --------------------------------------------------------------------------------
NeuralNetwork::NeuralNetwork(const std::size_t num_inputs, 
                             const std::size_t num_hidden_nodes, 
                             const std::size_t num_outputs,
                             const ActFunc act_func_hidden, 
                             const ActFunc act_func_output) 
    : hidden_layer_(DenseLayer(num_hidden_nodes, num_inputs, act_func_hidden))
    , output_layer_(DenseLayer(num_outputs, num_hidden_nodes, act_func_output)) {}

// --------------------------------------------------------------------------------
bool NeuralNetwork::AddTrainingData(const std::vector<std::vector<double>>& train_input,
                                    const std::vector<std::vector<double>>& train_output) {
    train_input_ = train_input; 
    train_output_ = train_output;
    CheckNumTrainingSets(); 
    InitTrainOrderVector(); 
    return NumTrainingSets() > 0;
}    

// --------------------------------------------------------------------------------
bool NeuralNetwork::Train(const std::size_t num_epochs, const double learning_rate) {
    if (NumTrainingSets() == 0 || num_epochs == 0 || learning_rate <= 0) { return false; }
    
    for (std::size_t i{}; i < num_epochs; ++i) {
        RandomizeTrainingOrder(); 
        for (const auto& j : train_order_) {
            Feedforward(train_input_[j]);
            Backpropagate(train_output_[j]);
            Optimize(train_input_[j], learning_rate);
        }
    }
    return true;
}

// --------------------------------------------------------------------------------
const std::vector<double>& NeuralNetwork::Predict(const std::vector<double>& input) {
     Feedforward(input);
     return output_layer_.Output(); 
}

// --------------------------------------------------------------------------------
void NeuralNetwork::PrintPredictions(const std::vector<std::vector<double>>& input_sets,
                                     const std::size_t num_decimals,
                                     std::ostream& ostream) {
    if (input_sets.size() == 0) { return; }
    ostream << std::fixed << std::setprecision(num_decimals);
    ostream << "--------------------------------------------------------------------------------";
    for (const auto& input: input_sets) {
        ostream << "\nInput:\t";
        Print<double>(input, ostream);
        ostream << "Output:\t";
        Print<double>(Predict(input), ostream);
    }
    ostream << "--------------------------------------------------------------------------------\n\n";
}

// --------------------------------------------------------------------------------
void NeuralNetwork::CheckNumTrainingSets() {
    if (train_input_.size() != train_output_.size()) {
        const auto num_sets{train_input_.size() < train_output_.size() ?
            train_input_.size() : train_output_.size()};
        train_input_.resize(num_sets);
        train_output_.resize(num_sets);
    }
}

// --------------------------------------------------------------------------------
void NeuralNetwork::InitTrainOrderVector() {
    train_order_.resize(train_input_.size());
    for (std::size_t i{}; i < train_order_.size(); ++i) {
        train_order_[i] = i; 
    }
}

// --------------------------------------------------------------------------------
void NeuralNetwork::RandomizeTrainingOrder() {
    utils::random::ShuffleVector<std::size_t>(train_order_);
}

// --------------------------------------------------------------------------------
void NeuralNetwork::Feedforward(const std::vector<double>& input) {
    hidden_layer_.Feedforward(input); 
    output_layer_.Feedforward(hidden_layer_.Output());
}

// --------------------------------------------------------------------------------
void NeuralNetwork::Backpropagate(const std::vector<double>& reference) {
    output_layer_.Backpropagate(reference);
    hidden_layer_.Backpropagate(output_layer_);
}

// --------------------------------------------------------------------------------
void NeuralNetwork::Optimize(const std::vector<double>& input, const double learning_rate) {
    hidden_layer_.Optimize(input, learning_rate);
    output_layer_.Optimize(hidden_layer_.Output(), learning_rate);
}

} /* namespace machine_learning */
} /* namespace yrgo */